File: src/repotask/index/runner.py
"""Walk the project and extract declarations into a symbol index."""

from __future__ import annotations

import fnmatch
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Any

from repotask.config.models import RepoTaskConfig
from repotask.index.languages import DECLARATIONS, language_for, parser_for
from repotask.output import warn

IDENTIFIER_NODES = {
    "identifier",
    "type_identifier",
    "simple_identifier",
    "field_identifier",
    "property_identifier",
}


@dataclass(frozen=True)
class Symbol:
    name: str
    kind: str
    language: str
    path: str
    line: int
    scope: str = ""


@dataclass
class IndexResult:
    symbols: list[Symbol]
    files_scanned: int
    files_skipped: int
    languages: dict[str, int]

    def as_dict(self) -> dict[str, Any]:
        return {
            "symbols": [asdict(symbol) for symbol in self.symbols],
            "filesScanned": self.files_scanned,
            "filesSkipped": self.files_skipped,
            "languages": self.languages,
        }


def build_index(config: RepoTaskConfig, paths: list[str] | None = None) -> IndexResult:
    """Index the whole project, or only `paths` when given (incremental refresh)."""
    files = _candidate_files(config, paths)
    symbols: list[Symbol] = []
    languages: dict[str, int] = {}
    scanned = 0
    skipped = 0
    missing_grammars: set[str] = set()

    for file in files:
        language = language_for(file.suffix)
        if language is None:
            continue
        if file.stat().st_size > config.index.max_file_bytes:
            skipped += 1
            continue
        parser = parser_for(language)
        if parser is None:
            missing_grammars.add(language)
            skipped += 1
            continue
        relative = str(file.relative_to(config.root))
        found = _extract(parser, language, file.read_bytes(), relative)
        symbols.extend(found)
        scanned += 1
        languages[language] = languages.get(language, 0) + 1

    for language in sorted(missing_grammars):
        warn(f"No tree-sitter grammar available for {language}; those files were skipped.")

    symbols.sort(key=lambda symbol: (symbol.path, symbol.line, symbol.name))
    return IndexResult(symbols, scanned, skipped, languages)


def _candidate_files(config: RepoTaskConfig, paths: list[str] | None) -> list[Path]:
    if paths is not None:
        candidates = [config.root / path for path in paths]
        return [path for path in candidates if path.is_file() and not _excluded(config, path)]
    return [
        path
        for path in sorted(config.root.rglob("*"))
        if path.is_file()
        and path.suffix.lower() in _known_suffixes()
        and not _excluded(config, path)
    ]


def _known_suffixes() -> set[str]:
    from repotask.index.languages import EXTENSIONS

    return set(EXTENSIONS)


def _excluded(config: RepoTaskConfig, path: Path) -> bool:
    relative = str(path.relative_to(config.root))
    if any(part.startswith(".") for part in Path(relative).parts[:-1]):
        return True
    return any(fnmatch.fnmatch(relative, pattern) for pattern in config.index.exclude)


def _extract(parser: Any, language: str, source: bytes, relative: str) -> list[Symbol]:
    declarations = DECLARATIONS.get(language, {})
    if not declarations:
        return []
    tree = parser.parse(source)
    symbols: list[Symbol] = []
    stack: list[tuple[Any, str]] = [(tree.root_node, "")]
    while stack:
        node, scope = stack.pop()
        child_scope = scope
        kind = declarations.get(node.type)
        if kind is not None:
            name = _name_of(node, source)
            if name:
                symbols.append(
                    Symbol(
                        name=name,
                        kind=kind,
                        language=language,
                        path=relative,
                        line=node.start_point[0] + 1,
                        scope=scope,
                    )
                )
                child_scope = f"{scope}.{name}" if scope else name
        for child in reversed(node.children):
            stack.append((child, child_scope))
    return symbols


def _name_of(node: Any, source: bytes) -> str:
    field = node.child_by_field_name("name")
    if field is None:
        # Not every grammar exposes a `name` field — Kotlin uses `simple_identifier`,
        # Dart signatures and Go methods hide the identifier one level down.
        field = next(
            (child for child in node.children if child.type in IDENTIFIER_NODES),
            None,
        )
    if field is None:
        return ""
    return source[field.start_byte : field.end_byte].decode("utf-8", errors="replace")

File: src/index/runner.rs
//! Walk the project and extract declarations into a symbol index.

use std::collections::BTreeMap;
use std::path::{Path, PathBuf};

use anyhow::Result;
use globset::{Glob, GlobSet, GlobSetBuilder};
use serde::Serialize;
use tree_sitter::{Node, Parser};

use crate::config::RepoTaskConfig;
use crate::index::languages::{
    declarations, grammar, is_known_suffix, language_for, IDENTIFIER_NODES,
};
use crate::output;

#[derive(Debug, Clone, Serialize)]
pub struct Symbol {
    pub name: String,
    pub kind: String,
    pub language: String,
    pub path: String,
    pub line: usize,
    pub scope: String,
}

#[derive(Debug, Default)]
pub struct IndexResult {
    pub symbols: Vec<Symbol>,
    pub files_scanned: usize,
    pub files_skipped: usize,
    pub languages: BTreeMap<String, usize>,
}

/// Index the whole project, or only `paths` when given (incremental refresh).
pub fn build_index(config: &RepoTaskConfig, paths: Option<&[String]>) -> Result<IndexResult> {
    let excludes = build_globset(&config.index.exclude);
    let files = candidate_files(config, paths, &excludes);

    let mut result = IndexResult::default();
    let mut missing: Vec<String> = Vec::new();
    let mut parser = Parser::new();

    for file in files {
        let Some(suffix) = file.extension().and_then(|value| value.to_str()) else {
            continue;
        };
        let Some(language) = language_for(suffix) else {
            continue;
        };
        let Ok(metadata) = file.metadata() else {
            result.files_skipped += 1;
            continue;
        };
        if metadata.len() > config.index.max_file_bytes {
            result.files_skipped += 1;
            continue;
        }
        let Some(grammar) = grammar(language) else {
            if !missing.iter().any(|item| item == language) {
                missing.push(language.to_string());
            }
            result.files_skipped += 1;
            continue;
        };
        if parser.set_language(&grammar).is_err() {
            result.files_skipped += 1;
            continue;
        }
        let Ok(source) = std::fs::read(&file) else {
            result.files_skipped += 1;
            continue;
        };
        let relative = file
            .strip_prefix(&config.root)
            .unwrap_or(&file)
            .to_string_lossy()
            .replace('\\', "/");

        result
            .symbols
            .extend(extract(&mut parser, language, &source, &relative));
        result.files_scanned += 1;
        *result.languages.entry(language.to_string()).or_insert(0) += 1;
    }

    missing.sort();
    for language in missing {
        output::warn(format!(
            "This build has no grammar for {language}; those files were skipped."
        ));
    }

    result.symbols.sort_by(|left, right| {
        left.path
            .cmp(&right.path)
            .then_with(|| left.line.cmp(&right.line))
            .then_with(|| left.name.cmp(&right.name))
    });
    Ok(result)
}

fn build_globset(patterns: &[String]) -> GlobSet {
    let mut builder = GlobSetBuilder::new();
    for pattern in patterns {
        if let Ok(glob) = Glob::new(pattern) {
            builder.add(glob);
        }
    }
    builder.build().unwrap_or_else(|_| GlobSet::empty())
}

fn candidate_files(
    config: &RepoTaskConfig,
    paths: Option<&[String]>,
    excludes: &GlobSet,
) -> Vec<PathBuf> {
    if let Some(paths) = paths {
        return paths
            .iter()
            .map(|path| config.root.join(path))
            .filter(|path| path.is_file() && !excluded(config, path, excludes))
            .collect();
    }
    let mut files: Vec<PathBuf> = walkdir::WalkDir::new(&config.root)
        .into_iter()
        .filter_entry(|entry| {
            // Never descend into dot directories; .git alone would dominate the walk.
            !entry.file_name().to_string_lossy().starts_with('.') || entry.depth() == 0
        })
        .flatten()
        .map(|entry| entry.into_path())
        .filter(|path| {
            path.is_file()
                && path
                    .extension()
                    .and_then(|value| value.to_str())
                    .is_some_and(is_known_suffix)
                && !excluded(config, path, excludes)
        })
        .collect();
    files.sort();
    files
}

fn excluded(config: &RepoTaskConfig, path: &Path, excludes: &GlobSet) -> bool {
    let Ok(relative) = path.strip_prefix(&config.root) else {
        return true;
    };
    let mut components: Vec<_> = relative.components().collect();
    components.pop();
    if components
        .iter()
        .any(|component| component.as_os_str().to_string_lossy().starts_with('.'))
    {
        return true;
    }
    excludes.is_match(relative.to_string_lossy().replace('\\', "/").as_str())
}

fn extract(parser: &mut Parser, language: &str, source: &[u8], relative: &str) -> Vec<Symbol> {
    let mapping = declarations(language);
    if mapping.is_empty() {
        return Vec::new();
    }
    let Some(tree) = parser.parse(source, None) else {
        return Vec::new();
    };

    let mut symbols: Vec<Symbol> = Vec::new();
    let mut stack: Vec<(Node, String)> = vec![(tree.root_node(), String::new())];
    while let Some((node, scope)) = stack.pop() {
        let mut child_scope = scope.clone();
        if let Some((_, kind)) = mapping
            .iter()
            .find(|(node_type, _)| *node_type == node.kind())
        {
            if let Some(name) = name_of(&node, source) {
                let kind = refine_kind(language, &node, kind);
                symbols.push(Symbol {
                    name: name.clone(),
                    kind: kind.to_string(),
                    language: language.to_string(),
                    path: relative.to_string(),
                    line: node.start_position().row + 1,
                    scope: scope.clone(),
                });
                child_scope = if scope.is_empty() {
                    name
                } else {
                    format!("{scope}.{name}")
                };
            }
        }
        let mut cursor = node.walk();
        let children: Vec<Node> = node.children(&mut cursor).collect();
        for child in children.into_iter().rev() {
            stack.push((child, child_scope.clone()));
        }
    }
    symbols
}

/// Swift folds `struct`, `class`, `actor`, and `enum` into a single
/// `class_declaration`; the leading keyword is what distinguishes them. Split
/// out `struct` so SwiftUI views and components can be targeted precisely.
/// Other keywords keep the mapped kind.
fn refine_kind<'a>(language: &str, node: &Node, kind: &'a str) -> &'a str {
    if language == "swift" && node.kind() == "class_declaration" {
        let mut cursor = node.walk();
        if node
            .children(&mut cursor)
            .any(|child| child.kind() == "struct")
        {
            return "struct";
        }
    }
    kind
}

fn name_of(node: &Node, source: &[u8]) -> Option<String> {
    let field = match node.child_by_field_name("name") {
        Some(field) => field,
        None => {
            // Collect first: the iterator borrows the cursor, but Node is tied to
            // the tree's lifetime, so the nodes outlive the walk.
            let mut cursor = node.walk();
            let children: Vec<Node> = node.children(&mut cursor).collect();
            children
                .into_iter()
                .find(|child| IDENTIFIER_NODES.contains(&child.kind()))?
        }
    };
    let text = source.get(field.start_byte()..field.end_byte())?;
    let name = String::from_utf8_lossy(text).to_string();
    (!name.is_empty()).then_some(name)
}
